ui/src/hooks/use-vulnerabilities-by-product-suspense.ts

```diff
@@ -0,0 +1,167 @@
+            run.jobs.advisor.finishedAt
+          ) {
+            return { ...run, repositoryId: repository.id };
+          }
+        }
+        return null;
+      },
+    })),
+  });
+
+  // Safety measure to ensure that only valid data is included in the array.
+  const ortRuns = useMemo(
+    () => runsForAdvisorQueries.filter((query) => query.data).filter(Boolean),
+    [runsForAdvisorQueries]
+  );
+
+  // Get the vulnerabilities found from the newest succeeded ORT runs.
+  const vulnerabilitiesQueries = useSuspenseQueries({
+    queries: ortRuns.flatMap((lastRunForRepository) => {
+      const lastRunData = lastRunForRepository.data;
+      return lastRunData
+        ? [
+            {
+              queryKey: [
+                'product',
+                ...UseVulnerabilitiesServiceGetVulnerabilitiesByRunIdKeyFn({
+                  runId: lastRunData.id,
+                  limit: 100000,
+                }),
+              ],
+              queryFn: async () => {
+                const vulnerabilities =
+                  await VulnerabilitiesService.getVulnerabilitiesByRunId({
+                    runId: lastRunData.id,
+                    limit: 100000,
+                  });
+
+                return vulnerabilities.data.map((vulnerability) => ({
+                  ...vulnerability,
+                  repositoryId: lastRunData.repositoryId,
+                }));
+              },
+            },
+          ]
+        : [];
+    }),
+  });
+
+  // Safety measure to ensure that only valid data is included in the array.
+  const vulnerabilities = useMemo(
+    () => vulnerabilitiesQueries.flatMap((query) => query.data).filter(Boolean),
+    [vulnerabilitiesQueries]
+  );
+
+  // Flatten the vulnerabilities and count occurrences by externalId and repositoryId.
+
+  const vulnerabilityRepoMap = useMemo(() => {
+    const map: Record<
+      string,
+      { vulnerability: VulnerabilityWithIdentifier; repoSet: Set<number> }
+    > = {};
+    vulnerabilities
+      .filter(
+        (
+          vulnerability
+        ): vulnerability is VulnerabilityWithIdentifier & {
+          repositoryId: number;
+        } => vulnerability !== undefined
+      )
+      .forEach(
+        (
+          vulnerability: VulnerabilityWithIdentifier & { repositoryId: number }
+        ) => {
+          const externalId = vulnerability.vulnerability.externalId;
+          if (!map[externalId]) {
+            map[externalId] = {
+              vulnerability,
+              repoSet: new Set(),
+            };
+          }
+          map[externalId].repoSet.add(vulnerability.repositoryId);
+        }
+      );
+    return map;
+  }, [vulnerabilities]);
+
+  // Convert the counts to an array of objects.
+  const result = useMemo(() => {
+    return Object.entries(vulnerabilityRepoMap).map(
+      ([, { vulnerability, repoSet }]) => ({
+        ...vulnerability,
+        repositoryCount: repoSet.size,
+      })
+    );
+  }, [vulnerabilityRepoMap]);
```

Comment:
I don't think there's need to have 3 useMemos, one should be enough that includes all the processing.